File: concepts/10_time_locale_and_runtime/02_calendar_time_zones_and_arithmetic/lua/test_02_transition_and_arithmetic_boundaries.lua
-- Common question: how do calendar arithmetic and zone transitions handle invalid/local times?
-- Inputs: normalized out-of-range fields, leap day, elapsed-second addition, and absent zone APIs.
-- Observations: os.time normalization, leap-year behavior, instant arithmetic, and transition-data absence.
-- polyglot-family: time_locale_and_runtime
-- polyglot-concept: calendar_time_zones_and_arithmetic
-- polyglot-related: languages/lua/standard_library/11_io_os_and_processes/
-- polyglot-related+: test_086_calendar_conversion_and_platform_boundaries.lua

local t = require("support.assertions")

local normalized = {year = 2026, month = 1, day = 32, hour = 0}
local timestamp = os.time(normalized)
t.equal(os.date("*t", timestamp).month, 2)
t.equal(os.date("*t", timestamp).day, 1)

local leap = os.time({year = 2024, month = 2, day = 29, hour = 12})
local leap_parts = os.date("*t", leap)
t.equal(leap_parts.month, 2)
t.equal(leap_parts.day, 29)

t.equal(rawget(os, "timezone"), nil)
t.equal(rawget(os, "zoneinfo"), nil)

-- 日历字段归一化不等于具备 DST gap/fold 规则；标准库没有 named-zone arithmetic。
t.done()

File: concepts/10_time_locale_and_runtime/02_calendar_time_zones_and_arithmetic/ruby/test_02_transition_and_arithmetic_boundaries.rb
# 共同问题：calendar arithmetic、无效日期和 zone transition 数据怎样处理。
# 输入：月底、闰日、elapsed seconds、Date invalid value 和固定 offset；观察：月份裁剪、错误与 instant arithmetic。
# polyglot-family: time_locale_and_runtime
# polyglot-concept: calendar_time_zones_and_arithmetic
# polyglot-related: languages/ruby/standard_library/10_data_time_and_text/test_075_date_time_and_calendar_arithmetic.rb

require "assertions"
require "date"

A = PolyglotAssertions

A.equal(Date.new(2026, 2, 28), Date.new(2026, 1, 31) >> 1)
A.equal(Date.new(2024, 2, 29), Date.new(2024, 2, 28) + 1)
A.raises(Date::Error) { Date.new(2026, 2, 30) }

instant = Time.new(2026, 1, 1, 0, 0, 0, "+05:30")
A.equal(19_800, instant.utc_offset)
A.equal(60.0, (instant + 60) - instant)
A.equal("+0530", instant.strftime("%z"))

# Core Time 接受 fixed offset；IANA transition database 不是 Ruby 语言接口。
A.falsey(Time.respond_to?(:zoneinfo))

A.done
